app/src/main/java/com/vibro/testviewpager2/PdfRenderingEngine.kt

```diff
@@ -0,0 +1,291 @@
+                        val height = if (isPortrait) pageSize.second else pageSize.first
+                        PageToSave(counter++, it.bitmap, width,height)
+                    }
+            } else {
+                Observable.fromCallable {
+                    val bitmap = bitmapFromUri(page.pageChangesData.storedPage) ?: throw IllegalArgumentException()
+                    PageToSave(counter++, bitmap, bitmap.width, bitmap.height)
+                }
+            }
+        }
+
+        fun writePageToPdfFile(pageToSave: PageToSave, pdfDocument: PdfDocument) {
+            val pageInfo = PdfDocument.PageInfo.Builder(pageToSave.width, pageToSave.height, pageToSave.index).create()
+            val page = pdfDocument.startPage(pageInfo)
+            page.canvas.drawBitmap(pageToSave.bitmap, null, Rect(0, 0, pageToSave.width, pageToSave.height), Paint())
+            pdfDocument.finishPage(page)
+        }
+
+        fun writeOnDisk(pdfDocument: PdfDocument): String {
+            val file = snRenderer.getCurrentFile().file
+            val path = file.absolutePath
+            file.delete()
+            val outStream = FileOutputStream(path)
+            pdfDocument.writeTo(outStream)
+            try {
+                outStream.close()
+            } catch (e: Exception) {
+                //do nothing
+            }
+            return path
+        }
+
+        return Observable.fromIterable(currentPages)
+            .flatMap { page -> providePageToSave(page) }
+            .map { pageToSave -> writePageToPdfFile(pageToSave, pdfDocument) }
+            .toList()
+            .toObservable()
+            .map { writeOnDisk(pdfDocument) }
+            .doFinally { pdfDocument.close() }
+    }
+
+    override fun getPages() = if (!hasChanges) originalPages else currentPages
+
+    override fun providePage(index: Int, quality: SnRenderer.Quality): Observable<RenderedPageData> {
+        if (!pageIsNewOrHasChanges(index)) {
+            val pageInfo = getPages()[index]
+            return snRenderer.renderPage(pageInfo, quality)
+        } else {
+            return Observable.fromCallable {
+                val currentPage = currentPages[index]
+                val b = bitmapFromUri(currentPage.pageChangesData?.storedPage)
+                RenderedPageData(currentPage, quality, b, RenderingStatus.Complete)
+            }
+        }
+    }
+
+    private fun bitmapFromUri(uri: Uri?): Bitmap? {
+        return Glide.with(context).asBitmap().load(uri)
+            .submit()
+            .get()
+    }
+
+    fun addPage(uri: Uri): Observable<Boolean> {
+        return Observable.fromCallable {
+            val pageIndex = currentPages.lastIndex + 1
+            val pageInfo = PageInfo(null, -1, pageIndex, null, PageChangesData(uri))
+            currentPages.add(pageInfo)
+            true
+        }
+    }
+
+    fun rearrangePage(indexFrom: Int, indexTo: Int): Observable<Unit> {
+        return Observable.fromCallable {
```

Comment:
We will need to change this method. It should work with arrays because the user can make several page rearrangement at once